_source/_posts/2023-06-22-net-maui-authentication.md

```diff
@@ -0,0 +1,582 @@
+
+            using (var httpClient = new HttpClient())
+            {
+                var discoveryDocumentResponse = await httpClient.GetDiscoveryDocumentAsync(
+                    new DiscoveryDocumentRequest
+                    {
+                        Address = Configuration.OktaDomain,
+                        Policy =
+                        {
+                            // Okta Org AS must be included to avoid https://stackoverflow.com/questions/56459997/endpoint-belongs-to-different-authority
+                            AdditionalEndpointBaseAddresses = new List<string> { oktaOrgAuthorizationServer }
+                        }
+
+                    }, cancellationToken).ConfigureAwait(false);
+
+                if (discoveryDocumentResponse.IsError)
+                {
+                    throw new InvalidOperationException("Error loading discovery document: " +
+                                                        discoveryDocumentResponse.Error, discoveryDocumentResponse.Exception);
+                }
+
+                _oidcClient.Options.ProviderInformation = new ProviderInformation
+                {
+                    IssuerName = discoveryDocumentResponse.Issuer,
+                    KeySet = discoveryDocumentResponse.KeySet,
+                    AuthorizeEndpoint = discoveryDocumentResponse.AuthorizeEndpoint,
+                    TokenEndpoint = discoveryDocumentResponse.TokenEndpoint,
+                    EndSessionEndpoint = discoveryDocumentResponse.EndSessionEndpoint,
+                    UserInfoEndpoint = discoveryDocumentResponse.UserInfoEndpoint,
+                    TokenEndPointAuthenticationMethods =
+                        discoveryDocumentResponse.TokenEndpointAuthenticationMethodsSupported,
+
+                };
+            }
+        }
+    }
+}
+```
+
+Before jumping into the `OktaClientConfiguration` creation, I'd like to explain a few differences between this implementation and the Auth0 client described in [Add Authentication to .NET MAUI Apps with Auth0](https://auth0.com/blog/add-authentication-to-dotnet-maui-apps-with-auth0/) blog post from Auth0 by Okta.
+
+Okta provides [two different types of authorization servers](https://developer.okta.com/docs/concepts/auth-servers/), the "Org Authorization Server" (Org AS) and the "Custom Authorization Server" (Custom AS). Suppose you try to use a Custom AS with IdentityModel. In that case. In that case, you will face the following error "[Error loading discovery document: Endpoint belongs to different authority: https://xxxxxxxxx.com/oauth2/v1/clients](https://devforum.okta.com/t/error-loading-discovery-document-endpoint-belongs-to-different-authority/6582)". This is because, during discovery, IdentityModel validates that all endpoints belong to the same authorization server, and with Okta, the client registration occurs at the Org level, not the authorization server level.
```

Comment:
```suggestion
Okta provides [two different types of authorization servers](https://developer.okta.com/docs/concepts/auth-servers/), the "Org Authorization Server" (Org AS) and the "Custom Authorization Server" (Custom AS). Suppose you try to use a Custom AS with IdentityModel. In that case, you will face the following error: "[Error loading discovery document: Endpoint belongs to different authority: https://xxxxxxxxx.com/oauth2/v1/clients](https://devforum.okta.com/t/error-loading-discovery-document-endpoint-belongs-to-different-authority/6582)". This is because during discovery, IdentityModel validates that all endpoints belong to the same authorization server, and with Okta, the client registration occurs at the Org level, not the authorization server level.
```